proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,280 @@
+
+## New Terminology
+
+* UserBlock - a block full of non vote transactions.
+
+* UserBlockEntry - this is the entry that leader creates in PoH for
+the UserBlock, it contains a hash of the built UserBlock.
+
+* UserBlockSlot - A slot for builders to transmit the UserBlock to
+the cluster over turbine. Cluster could be configured with more
+then one UserBlockSlots per slot with a default setting of 2.
+
+* Leader - the current leader for the slot that will build a PoH
+ledger full of Votes and UserBlockEntries.
+
+* Builder - a node that is scheduled to build a block with non
+vote transactions
+
+* VoteHash - like BankHash, but for vote only transactions.
+
+* BundleTransactions - a transaction signed by the builder that can
+list transactions in the builders UserBlock to be executed in
+one ordered batch. This transaction can also add a priority fee
+such that the entire bundle is prioritized for execution as a batch.
+This transaction can only appear in the builders UserBlock and can
+only reference transactions prior to it in the UserBlock. TBD on
+format.
+
+## Detailed Design
+
+### Overview
+Leaders are scheduled to build blocks as they are currently by
+the LeaderSchedule.
+
+Builder's are scheduled along side leaders to build UserBlocks -
+blocks of non vote transactions. N number of builders can be scheudled
+concurrently to build blocks.
+
+While a leader is scheduled, they receive and encode votes as normal.
+Any well formed UserBlocks that were received from the previous or
+current UserBlockSlot are added to the leaders PoH ledger as
+UserBlockEntry.
+
+The N concurrent Builder create blocks out of user transactions.
+These are transmitted to the cluster via turbine.  The leader
+receives and decodes them and generates a UserBlockEntry, and adds
+it to PoH as soon as the leaders PoH has started the UserBlockSlot.
+
+Validators can vote on leader blocks by executing the votes, but
+before completing the execution of UserBlock transactions.
+
+### Fork Choice
+
+If a validator doesn't have the builder's UserBlock, the validator
+doesn't vote on the proposed block and tries to repair the UserBlock.
+That fork is ignored for fork choice.
+
+Otherwise the validator evaluates the forks as usual. 
+
+### UserBlock schedule
+
+There are 2 UserBlockSlots per slot. Builders produce 1 UserBlock
+per UserBlockSlot. So if the cluster is configured with 3 Builders,
+each leader slot would see a total of 6 UserBlocks per slot.  3 in
+the first half of the leaders slot, and 3 in the second half.
+
+The current leader schedule rotates each leader every 4 slots, but
+builders could be rotated at a different schedule then the leaders
+and rotation could be staggered.
+
+### UserBlock Compute Limits
+
+If the overall compute capacity for user transactions per leader
+block is 48m CU, and cluster is configured with 2 builders, then
+each UserBlock can use no more then 48m/4 or 12m CU.
```

Comment:
This is a reasonable place to start imo. However, I have concerns that this could have unintentional side effects that lead to smaller blocks.

Validators are **not** spread evenly across the world, nor are users. There is significant clustering in North America and Europe.

As a hypothetical, let's say we have 2 builders 1 in Amsterdam and the other in Wellington NZ (roughly opposite sides of the world). The majority of users would simply target their nearest builder which will likely be Amsterdam. Amsterdam block gets packed and leaves people out while the Wellington block sits (relatively) idle.

It's certainly not a liveness issue, but I wonder if there is something that could be done to reduce this potential impact.

-------------------------------------------
**NOT WELL THOUGHT OUT SPITBALL IDEA**

Could we have more builders than can possibly fit into the `UserBlockSlot`, and leader takes the most heavily packed `UserBlockEntries`. This might be more complicated since then the leader has to scan these entries before selecting it.

Example above: 3 builders (AMS, WEL, +TOR).
AMS, TOR get packed because they are geographically (reduced latency) closer to users, and WEL gets excluded by the leader.


ugh, thinking about this more...maybe not a good idea. There's no way for non-leader to verify the leader chose the most packed block. Though the leader would (hopefully) be incentivized to choose the most packed due to fee-collection?

Reply:
Yea, I think this is a v1 optimization.

Reply:
Another reason for probably very small blocks: the builder only knows the compute limits requested not the compute units consumed.